Fix #354 - Incorrect setting the expiration date for a cookies

danff opened the pull request:
Fix for #354 
- Updated PsCookie, RsFlash and RsReturn classes to substract timezone offset from expire date when creating the cookie;
- Added to PsCookieTest, RsFlashTest and RsReturnTest classes test methods for testing the correct expire value to be in GMT timezone.

davvd commented:
@danff Thanks, I will find someone to review this PR soon

davvd commented:
@dmzaytsev please review, thanks

dmzaytsev commented on src/main/java/org/takes/facets/auth/PsCookie.java:
```diff
@@ -124,8 +126,8 @@ public Response exit(final Response res,
                 new Date(
-                    System.currentTimeMillis()
-                        + TimeUnit.DAYS.toMillis(this.age)
+                    millis + TimeUnit.DAYS.toMillis(this.age)
```
@danff why you dont want to use same method as in tests ?

it can be a constant
```
final SimpleDateFormat format = new SimpleDateFormat(
            "'Expires='EEE, dd MMM yyyy HH:mm:ss 'GMT'"
            Locale.ENGLISH
        );
```
and set the time zone in a static initializer
```
format.setTimeZone(TimeZone.getTimeZone("GMT"));
```

so we can remove once used variable `millis` and make the expression simplier

dmzaytsev commented on src/main/java/org/takes/facets/flash/RsFlash.java:
```diff
@@ -177,8 +179,8 @@ private static Response make(final String msg, final Level level,
                 new Date(
-                    System.currentTimeMillis()
-                        + TimeUnit.HOURS.toMillis(1L)
+                    millis + TimeUnit.HOURS.toMillis(1L)
```
@danff see comment in `PsCookie.java`

dmzaytsev commented on src/main/java/org/takes/facets/ret/RsReturn.java:
```diff
@@ -75,6 +76,9 @@ public RsReturn(final Response res, final String loc, final String cookie)
                     new Date(
                         System.currentTimeMillis()
                             + TimeUnit.HOURS.toMillis(1L)
+                            - TimeZone.getDefault().getOffset(
```
@danff see comment in `PsCookie.java`

dmzaytsev commented on src/test/java/org/takes/facets/auth/PsCookieTest.java:
```diff
@@ -49,11 +59,42 @@ public void addsCookieToResponse() throws IOException {
                 new PsCookie(
                     new CcPlain(), "foo", 1L
-                ).exit(new RsEmpty(), new Identity.Simple("urn:test:99"))
+                ).exit(new RsEmpty(), PsCookieTest.identity)
```
@danff let's use different input data for each test
and remove the constant

dmzaytsev commented on src/test/java/org/takes/facets/auth/PsCookieTest.java:
```diff
@@ -49,11 +59,42 @@ public void addsCookieToResponse() throws IOException {
     }
+
+    /**
+     * The expiration date for a cookie must be a GMT time.
```
@danff please check put this article about test method names and javadoc 
www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html?expand=1#test-method-names

dmzaytsev commented on src/test/java/org/takes/facets/flash/RsFlashTest.java:
```diff
@@ -63,4 +68,32 @@ public void addsCookieToResponse() throws IOException {
     }
+
+    /**
+     * The expiration date for a cookie must be a GMT time.
```
@danff please check put this article about test method names and javadoc 
www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html?expand=1#test-method-names

dmzaytsev commented on src/test/java/org/takes/facets/ret/RsReturnTest.java:
```diff
@@ -62,4 +67,32 @@ public void addsCookieToResponse() throws IOException {
     }
+
+    /**
+     * The expiration date for a cookie must be a GMT time.
```
@danff please check put this article about test method names and javadoc 
www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html?expand=1#test-method-names

dmzaytsev commented:
@danff please see my comments

dmzaytsev commented:
@danff btw see https://github.com/yegor256/takes/issues/354#issuecomment-116428216

> yegor256 commented on 29 Jun 2015
 and it's an obvious code duplication, we should resolve it, you're right

I would suggest to move the cookie expiration formatting to `RsWithCookie`

dmzaytsev commented:
@danff one more constructor for `RswithCookie` with a parameter  `Opt<Date>  expiries`

@yegor256 what do you think?

danff commented:
@dmzaytsev Added two more constructors and moved the expires formatting code to RsWithCookie class as requested. Fixed the test method names and javadocs.

dmzaytsev commented on src/main/java/org/takes/rs/RsWithCookie.java:
```diff
@@ -90,6 +95,18 @@
+     * Pattern for the expires cookie attribute.
+     */
+    private static final SimpleDateFormat EXPIRES_FORMAT = new SimpleDateFormat(
+        "'Expires='EEE, dd MMM yyyy HH:mm:ss 'GMT'",
```
@danff I think  it would better to use `'Expires='EEE, dd MMM yyyy HH:mm:ss z` instead

dmzaytsev commented on src/test/java/org/takes/facets/auth/PsCookieTest.java:
```diff
@@ -40,6 +45,11 @@
     /**
+     * Identity instance used in tests.
+     */
+    private static Identity identity = new Identity.Simple("urn:test:99");
```
@danff let's get rid the constant and use different identities for each test

dmzaytsev commented on src/test/java/org/takes/facets/flash/RsFlashTest.java:
```diff
@@ -63,4 +68,32 @@ public void addsCookieToResponse() throws IOException {
+     * @throws IOException If there are some problems inside
+     */
+    @Test
+    public void createCookieWithExpiresInGMT() throws IOException {
```
@danff `createsCookie...`
same for other tests

dmzaytsev commented:
@danff just a few minor comments

dmzaytsev commented:
@danff also please make sure Travis build without errors

danff commented:
@yegor256 @dmzaytsev FindBugs (and some javadocs) says DateFormat is not thread safe actually, so using a constant + static initializer for it is a bad idea. This leaves 2 options: first to create and initialize and instance of SimpleDateFormat in the RsWithCookie.make() method and leave the extra constructors or the second option to leave the initialization (and creation) of SimpleDateFormat in the PsCookie, RsFlash and RsReturn classes and leave RsWithCookie without those two extra constructors.

dmzaytsev commented:
@danff let's move it to `RsWithCookie.make() `

danff commented:
@dmzaytsev Moved the formatting code to make() method, fixed test methods.

dmzaytsev commented:
@danff thank you

dmzaytsev commented:
@rultor merge

rultor commented:
> @rultor merge

@dmzaytsev Thanks for your request. @yegor256 Please confirm this.

danff commented:
@dmzaytsev @yegor256 any progress here?

dmzaytsev commented:
@danff only @yegor256 can confirm the merge request as hi is the architect on this project
we should wait until Yegor return from a business trip

danff commented:
@dmzaytsev ok thanks

danff commented:
@yegor256 any progress with merge?

dmzaytsev commented:
@yegor256 ping

yegor256 commented:
@danff oh... it's a bad idea - to use `Opt` like that. just create different constructors. one with the date, another one without.

dmzaytsev commented:
@danff please close the pull request if you stopped working on the ticket

dmzaytsev commented:
@danff ping

dmzaytsev commented:
@yegor256 it seems @danff won't continue to work on it 
so could you please close this PR?

yegor256 commented:
@dmzaytsev closed

davvd commented:
@ypshenychka please, check this issue for QA compliance, as per [par.24](http://at.teamed.io/policy.html#24)

ypshenychka commented:
@davvd Quality is good.

davvd commented:
> @davvd Quality is good.

@ypshenychka thanks, got it, great to see high-quality tickets here

davvd commented:
@dmzaytsev just added 10 mins to @ypshenychka (for QA), payment ID is `83510775`

Thanks so much! Your account was topped up for **38 mins** (transaction ID is `83510796`, the task took 1673 hours and 1 min)

extra minutes for review comments (c=23)

+38 added to your rating, current score is: [+2619](http://www.netbout.com/b/35166?open=rating)
